File: FontPack/PackedFont.cs
// FontPacker 0.3 Beta Release
//
// Now works with any type of font listed in the font selected dialog
//
// Author: Sri Harsha Chilakapati

using System;
using System.IO;
using System.Xml;
using System.Text;
using System.Drawing;
using System.Collections.Generic;

namespace FontPack
{
    
    /// <summary>
    /// PackedFont - Class to read fonts packed by FontPacker
    /// </summary>
    public class PackedFont
    {
        
        Dictionary<char, Image> glyphs = new Dictionary<char, Image>();
        Dictionary<char, int> advances = new Dictionary<char, int>();

        /// <summary>
        /// Load a PackedFont from a file.
        /// </summary>
        /// <param name="filename">The name of the file</param>
        public PackedFont(string filename) : this(new FileStream(filename, FileMode.Open)) { }

        /// <summary>
        /// Load a PackedFont from a stream.
        /// </summary>
        /// <param name="stream">The stream containing the data.</param>
        public PackedFont(Stream stream)
        {
            XmlReader fontpack = XmlReader.Create(stream);

            // Read from the XML
            while (fontpack.Read())
            {
                switch (fontpack.NodeType)
                {
                    case XmlNodeType.Element:
                        // Read every glyph
                        if (fontpack.LocalName == "Glyph")
                        {
                            char glyph = fontpack.GetAttribute("char")[0];

                            byte[] image = Convert.FromBase64String(fontpack.GetAttribute("data"));

                            int xadvance = int.Parse(fontpack.GetAttribute("xadvance"));
                            advances.Add(glyph, xadvance);
                            
                            MemoryStream imgStream = new MemoryStream();

                            imgStream.Write(image, 0, image.Length);
                            imgStream.Seek(0, SeekOrigin.Begin);

                            Image bmp = Image.FromStream(imgStream);

                            glyphs.Add(glyph, bmp);
                        }
                        break;
                }
            }

            fontpack.Close();
            stream.Close();
        }
        
        /// <summary>
        /// Get the Glyph of the char
        /// </summary>
        /// <param name="c">The character of glyph</param>
        /// <returns>The Glyph as an Image</returns>
        public Image GetGlyph(char c)
        {
            return glyphs[c];
        }
        
        /// <summary>
        /// Get the AdvanceWidth of a glyph
        /// </summary>
        /// <param name="c">The character of glyph</param>
        /// <returns>The amount to move forward after drawing the glyph</returns>
        public int GetAdvanceWidth(char c)
        {
            return advances[c];
        }
        
        /// <summary>
        /// Gets the Dictionary containing the Glyphs
        /// </summary>
        public Dictionary<char, Image> Glyphs
        {
            get
            {
                return glyphs;
            }
        }
        
        /// <summary>
        /// Gets the Dictionary containing the AdvanceWidths
        /// </summary>
        public Dictionary<char, int> AdvanceWidths
        {
            get
            {
                return advances;
            }
        }

        /// <summary>
        /// Draws a string using this font at a position using the given Graphics context.
        /// </summary>
        /// <param name="text">The text to be drawn.</param>
        /// <param name="x">The x-position</param>
        /// <param name="y">The y-position</param>
        /// <param name="g">The graphics context to be used</param>
        public void DrawString(string text, float x, float y, Graphics g)
        {
            float posX = x;
            float posY = y;

            foreach (char ch in text)
            {
                if (ch == '\n')
                {
                    posY += glyphs[' '].Height;
                }
                else
                {
                    Image bmp = glyphs[ch];
                    g.DrawImageUnscaled(bmp, (int)posX, (int)posY);

                    posX += advances[ch];
                }
            }
        }

    }
}

File: JFontPack/com/fontpacker/fontpack/PackedFont.java
package com.fontpacker.fontpack;

import java.awt.Graphics2D;
import java.awt.Image;
import java.io.ByteArrayInputStream;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;

import javax.imageio.ImageIO;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;

import org.apache.commons.codec.binary.Base64;
import org.xml.sax.SAXException;
import org.xml.sax.Attributes;
import org.xml.sax.helpers.DefaultHandler;

/**
 * PackedFont - Class to read fonts packed with FontPacker
 * 
 * @author Sri Harsha Chilakapati
 */
public class PackedFont
{
    
    // The map of glyphs
    private HashMap<Character, Image> glyphs = new HashMap<Character, Image>();
    // The map of advance widths
    private HashMap<Character, Integer> advances = new HashMap<Character, Integer>();
    
    /**
     * Constructs a new PackedFont from the name of the file.
     * @param fileName The name of fontpack file
     * @throws FileNotFoundException If the file is not found.
     */
    public PackedFont(String fileName) throws FileNotFoundException
    {
        this (new FileInputStream(fileName));
    }
    
    /**
     * Constructs a new PackedFont from an InputStream.
     * @param stream The InputStream of the fontpack xml data.
     */
    public PackedFont(InputStream stream)
    {
        try
        {
            // Create a new SAX Parser
            SAXParser fontpack = SAXParserFactory.newInstance().newSAXParser();
            DefaultHandler handler = new DefaultHandler()
            {
                @Override
                public void startElement(String uri, String localName, String qName, Attributes attr)
                {
                    // Read every glyph from the font.
                    if (qName.equalsIgnoreCase("Glyph"))
                    {
                        try
                        {
                            // Get the character value
                            char ch = attr.getValue("char").charAt(0);
                            
                            // Get the advance width
                            int xadvance = Integer.parseInt(attr.getValue("xadvance"));
                            
                            // Decode the glyph image
                            String data = attr.getValue("data");
                            byte[] png = Base64.decodeBase64(data);
                            ByteArrayInputStream source = new ByteArrayInputStream(png);
                            Image img = ImageIO.read(source);
                            
                            // Map the Glyph and AdvanceWidth with the character
                            glyphs.put(ch, img);
                            advances.put(ch, xadvance);
                        }
                        catch (IOException e)
                        {
                            e.printStackTrace();
                        }
                    }
                }
            };
            
            // Parse the file.
            fontpack.parse(stream, handler);
            
            // Close the stream.
            stream.close();
        }
        catch (ParserConfigurationException | SAXException | IOException e)
        {
            e.printStackTrace();
        }
    }
    
    /**
     * @return The Map containing all the Glyphs
     */
    public HashMap<Character, Image> getGlyphs()
    {
        return glyphs;
    }
    
    /**
     * @return The Map containing all the AdvanceWidths
     */
    public HashMap<Character, Integer> getAdvanceWidths()
    {
        return advances;
    }
    
    /**
     * @param ch The Character value
     * @return The Glyph for the character
     */
    public Image getGlyph(char ch)
    {
        return glyphs.get(ch);
    }
    
    /**
     * @param ch The Character value
     * @return The amount to advance horizontally after drawing a glyph of this character
     */
    public int getAdvanceWidth(char ch)
    {
        return advances.get(ch);
    }
    
    /**
     * Draws a string of characters at a position on the screen using
     * a graphics context. Skips to newline on entering '\n' new line character.
     * @param text The text to be drawn.
     * @param posX The x-coordinate on the component
     * @param posY The y-coordinate on the component
     * @param g The graphics context.
     */
    public void drawString(String text, float posX, float posY, Graphics2D g)
    {
        float x = posX;
        float y = posY;
        
        for (char ch : text.toCharArray())
        {
            if (ch == '\r')
            {
                continue;
            }
            
            if (ch == '\n')
            {
                y += glyphs.get(' ').getHeight(null);
            }
            
            g.drawImage(glyphs.get(ch), (int)x, (int)y, null);
            x += advances.get(ch);
        }
    }

}
